runtime/include/chpl-mem-array.h

```diff
@@ -98,6 +98,69 @@ void chpl_mem_array_postAlloc(void* p, size_t nmemb, size_t eltSize,
 }
 
 
+static inline
+void* chpl_mem_array_realloc(void* p, size_t oldNmemb, size_t newNmemb,
+                             size_t eltSize,
+                             c_sublocid_t subloc, chpl_bool* callPostAlloc,
+                             int32_t lineno, int32_t filename) {
+  //
+  // Support for dynamic array registration by comm layers is done here
+  // via *callPostAlloc, the same as for chpl_mem_array_alloc().  See
+  // there for further information.
+
+  const size_t newSize = newNmemb * eltSize;
+  chpl_memhook_realloc_pre(p, newSize, CHPL_RT_MD_ARRAY_ELEMENTS,
+                           lineno, filename);
+
+  const size_t oldSize = oldNmemb * eltSize;
+  void* newp = NULL;
+  *callPostAlloc = false;
+  if (chpl_mem_size_justifies_comm_alloc(oldSize)) {
+    newp = chpl_comm_regMemRealloc(p, oldSize, newSize,
+                                   CHPL_RT_MD_ARRAY_ELEMENTS,
+                                   lineno, filename);
+    if (newp != NULL) {
+      *callPostAlloc = true;
+    }
+  }
+
+  if (newp == NULL) {
+    newp = chpl_realloc(p, newSize);
+  }
+
+  if (newp != p) {
+    //
+    // We allocated new memory.  Copy the existing bytes.  If we're on
+    // a NUMA compute node this will result in locality that matches
+    // that of the calling thread.  That's not very good but we cannot
+    // do better in the runtime.  At least with the copying here rather
+    // than in the comm layer it will be easier to hoist it into the
+    // module code if we can do better there someday.
+    //
+    memcpy(newp, p, oldSize);
```

Comment:
Since the typical behavior of `realloc` is to do the `memcpy()` if a new memory region was allocated, I'm surprised that this memcpy() would occur here rather than in the definitions of `chpl_realloc()` or `chpl_comm_regMemRealloc()` above.  Put another way, if we ended up calling C's realloc in the comm=none, mem=cstdlib configuration, wouldn't this result in an extra copy?  Or does that case go down a different code path that I'm not picking up on?

Reply:
Yep, that's a bug; I'll move the copy into the comm layer.